wdl-engine/src/config.rs

```diff
@@ -0,0 +1,221 @@
+//! Implementation of engine configuration.
+
+use std::sync::Arc;
+
+use anyhow::Context;
+use anyhow::Result;
+use anyhow::bail;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::SYSTEM;
+use crate::TaskExecutionBackend;
+use crate::convert_unit_string;
+use crate::local::LocalTaskExecutionBackend;
+
+/// Represents WDL evaluation configuration.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct Config {
+    /// Workflow evaluation configuration.
+    #[serde(default)]
+    pub workflow: WorkflowConfig,
+    /// Task evaluation configuration.
+    #[serde(default)]
+    pub task: TaskConfig,
+    /// Task execution backend configuration.
+    #[serde(default)]
+    pub backend: BackendConfig,
+}
+
+impl Config {
+    /// Validates the evaluation configuration.
+    pub fn validate(&self) -> Result<()> {
+        self.workflow.validate()?;
+        self.task.validate()?;
+        self.backend.validate()?;
+        Ok(())
+    }
+
+    /// Creates a new task execution backend based on this configuration.
+    pub fn create_backend(&self) -> Result<Arc<dyn TaskExecutionBackend>> {
+        match self.backend.default {
+            Backend::Local => Ok(Arc::new(LocalTaskExecutionBackend::new(
+                &self.backend.local,
+            )?)),
+        }
+    }
+}
+
+/// Represents workflow evaluation configuration.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct WorkflowConfig {
+    /// Scatter statement evaluation configuration.
+    #[serde(default)]
+    pub scatter: ScatterConfig,
+}
+
+impl WorkflowConfig {
+    /// Validates the workflow configuration.
+    pub fn validate(&self) -> Result<()> {
+        self.scatter.validate()?;
+        Ok(())
+    }
+}
+
+/// Represents scatter statement evaluation configuration.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct ScatterConfig {
+    /// The number of scatter array elements to process concurrently.
+    ///
+    /// By default, the value is the parallelism supported by the task
```

Comment:
I'm wondering (a) if this is necessary and (b) if it makes a difference in speed. If I'm understanding correctly (which I may not be), this would also limit the number of concurrent scatters that don't actually call a task, which seems a pity.

Reply:
I now see the other comment that explains this a bit. I think it's fine, though I am still curious about the speed (does it come out to roughly the same speed as executing them all concurrently?).

Reply:
The speed is roughly the same, but evaluation isn't usually the bottleneck for actually running a workflow, it's task execution since that will (soon) involve spinning up containers; evaluation will be dominated by waits.